pennylane/data/s3.py

```diff
@@ -0,0 +1,254 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the Dataset utility functions.
+"""
+# pylint:disable=too-many-arguments
+from concurrent.futures import ThreadPoolExecutor, wait
+import os
+import requests
+from pennylane.data.dataset import Dataset
+
+S3_URL = "https://xanadu-quantum-datasets-test.s3.amazonaws.com"
+FOLDERMAP_URL = os.path.join(S3_URL, "foldermap.json")
+DATA_STRUCT_URL = os.path.join(S3_URL, "data_struct.json")
+
+_foldermap = {}
+_data_struct = {}
+
+
+def _validate_params(data_type, description, attributes):
+    r"""Validate parameters for loading the data"""
+
+    data = _data_struct.get(data_type)
+    if not data:
+        raise ValueError(
+            f"Currently we have data hosted from types: {list(_data_struct)}, but got {data_type}."
+        )
+
+    params_needed = data["params"]
+    if set(description) != set(params_needed):
+        raise ValueError(
+            f"Supported parameter values for {data_type} are {params_needed}, but got {list(description)}."
+        )
+
+    def panic_if_invalid(node, params_left):
+        param = params_left[0]
+        params_left = params_left[1:]
+        detail = description[param]
+        if detail == "full":
+            if not params_left:
+                return
+            for child in node.values():
+                panic_if_invalid(child, params_left)
+        elif detail not in node:
+            raise ValueError(
+                f"{param} value of '{detail}' not available. Available values are {list(node)}"
+            )
+        elif params_left:
+            panic_if_invalid(node[detail], params_left)
+
+    panic_if_invalid(_foldermap[data_type], params_needed)
+
+    if not isinstance(attributes, list):
+        raise TypeError(f"Arg 'attributes' should be a list, but got {type(attributes)}.")
```

Comment:
it could be, decided offline to leave it as is. users can change it quickly. plus it might help in the future when adding things like `layout` being a tuple.